packages/http-client/src/csrf-token-middleware.ts

```diff
@@ -0,0 +1,199 @@
+import {
+  createLogger,
+  ErrorWithCause,
+  first,
+  flatten,
+  pickIgnoreCase,
+  pickValueIgnoreCase,
+  removeTrailingSlashes
+} from '@sap-cloud-sdk/util';
+import axios from 'axios';
+import { executeWithMiddleware } from '@sap-cloud-sdk/resilience/internal';
+import {
+  HttpMiddleware,
+  HttpMiddlewareContext,
+  HttpMiddlewareOptions,
+  HttpRequestConfig,
+  HttpRequestConfigBase,
+  HttpRequestConfigWithOrigin,
+  Method
+} from './http-client-types';
+
+const logger = createLogger('csrf-middleware');
+
+/**
+ * Options for middleware to fetch csrf tokens.
+ */
+export interface CsrfMiddlewareOptions {
+  /**
+   * Method used for the token fetching. Default is `head`.
+   */
+  method?: Method;
+  /**
+   * Url used for the token fetching. Default is the resource path without parameters.
+   */
+  url?: string;
+  /**
+   * Middlewares added to the token retrieval request.
+   */
+  middleware?: HttpMiddleware[];
+}
+
+/**
+ * Middleware for fetching a CSRF token. This middleware is added to all request per default.
+ * Use the `fetchCsrfToken` option to disable it.
+ * @param options - Options like URL or method to configure the token fetching.
+ * @returns The middleware for fetching CSRF tokens.
+ */
+export function csrf(options?: CsrfMiddlewareOptions): HttpMiddleware {
+  return (middlewareOptions: HttpMiddlewareOptions) => async requestConfig => {
+    if (noActionNeeded(requestConfig)) {
+      return middlewareOptions.fn(requestConfig);
+    }
+    const csrfToken = await makeCsrfRequests(requestConfig, {
+      ...options,
+      ...middlewareOptions
+    });
+    if (csrfToken?.cookie) {
+      csrfToken.cookie = requestConfig.headers?.cookie
+        ? [requestConfig.headers?.cookie, csrfToken?.cookie].join(';')
+        : csrfToken?.cookie;
+    }
+    requestConfig.headers = { ...requestConfig.headers, ...csrfToken };
+    return middlewareOptions.fn(requestConfig);
+  };
+}
+
+function noActionNeeded(requestConfig: HttpRequestConfig): boolean {
```

Comment:
[req] According to our style guide a function should start with a verb. So, this would probably be: `isNoActionNeeded`.
[pp] Action is very general and the negation in the name adds unnecessary potential for confusion in my opinion. How about returning the opposite boolean and calling it `needsCsrfToken`?